bwosqueue/src/lib.rs

```diff
@@ -0,0 +1,832 @@
+                let t =
+                    unsafe { blk.entries[reserved_idx].with(|entry| entry.read().assume_init()) };
+                // `t` is now owned by us so we mark the stealing as finished. Synchronizes with the Owner Acquire.
+                let old_stolen = blk.stolen.fetch_add(1, Release);
+                debug_assert!(old_stolen.raw_index() < ENTRIES_PER_BLOCK);
+                return Some(t);
+            }
+
+            // Slow-path: The current block is already fully reserved. Try to advance to the next block
+            if !self.can_advance(blk, reserved) {
+                return None;
+            }
+            self.try_advance_spos(curr_spos);
+        }
+    }
+
+    /// Get the current stealer `Block` and the corresponding stealer position (`spos`)
+    ///
+    /// The returned `spos` can be larger than `NUM_BLOCKS` to detect [ABA](https://en.wikipedia.org/wiki/ABA_problem)
+    /// situations.
+    fn curr_block(&self) -> (&Block<E, ENTRIES_PER_BLOCK>, usize) {
+        let curr_spos = self.spos.load(Relaxed);
+        // spos increments beyond NUM_BLOCKS to prevent ABA problems.
+        let block_idx = curr_spos % NUM_BLOCKS;
+        let blk: &Block<E, ENTRIES_PER_BLOCK> = &self.queue.blocks[block_idx];
+        (blk, curr_spos)
+    }
+
+    /// Try to steal a block from `self`.
+    ///
+    /// Tries to steal a full block from `self`. If the block is not fully
+    /// committed yet it will steal up to and including the last committed entry
+    /// of that block.
+    #[inline]
+    pub fn steal_block(&self) -> Option<StealerBlockIter<'_, E, ENTRIES_PER_BLOCK>> {
+        loop {
+            let (blk, curr_spos) = self.curr_block();
+
+            /* check if the block is fully reserved */
+            let reserved = blk.reserved.load(Acquire);
+            let reserved_idx = reserved.raw_index();
+
+            if reserved_idx < ENTRIES_PER_BLOCK {
+                /* check if we have an entry to occupy */
+                let committed = blk.committed.load(Acquire);
+                let committed_idx = committed.raw_index();
+                if reserved_idx == committed_idx {
+                    return None;
+                }
+
+                // Try to steal the block up to the latest committed entry
+                let reserve_res = blk
+                    .reserved
+                    .compare_exchange_weak(reserved, committed, Release, Relaxed);
+
+                if reserve_res.is_err() {
+                    return None;
+                }
+
+                let num_reserved = committed_idx - reserved_idx;
+                // From the statistics perspective we consider the reserved range to already be
+                // stolen, since it is not available for the consumer or other stealers anymore.
+                #[cfg(feature = "stats")]
+                self.queue.stats.increment_stolen(num_reserved);
```

Comment:
No, that would be wrong. At this point we have just reserved the entries, i.e. the consumer can't access them anymore, but the stealer has not finished copying the entries over to the stealers queue. The `Drop` implementation for `StealerBlockIter` increments `stolen` once the (now empty) iterator is dropped.

The statistics feature is just an approximation, giving metrics on how many items are in the queue, so that the utilization could be tracked over time.